### ISSUE TITLE
Docker architecture provides bad DX

### ISSUE THREAD

**Issue by MattGson** [cube-js/cube.js#1561]
TLDR: The new architecture feels more like a pre-made service with js syntax for configuration, not a js library. The problem is that it makes the use of existing or shared js code difficult or impossible.

**Describe the bug**

Based on the blog article, docs and new defaults, It looks like cube will remove support for using the CubejsServer. 

The new docker based architecture is very hard to work with for existing projects. Cube's reason for changing talks about express limitations but there is no reason cube can't cut out express while allowing the use of a js entrypoint like before.

The problem is not deploying cube in a docker container (which was already trivial), the problem is incorporating code that does not fit the expected structure. The recommended structure also enforces unconventional code such as installing production dependencies as dev-dependencies (makes intent unclear).

**To Reproduce**
Create a cubejs project where cube-server and connector are installed as actual dependencies (not dev-deps). Put the cube.ts file somewhere other than the root of the project (think mono-repo with many services). Add a custom docker image for your cubejs service that incorporates an existing re-usable build step.

- Try to run the dev server: Requires calling a node_module directly (Yuck). fails with error about dev-deps.
- Try to run the prod docker image: fails because cube-cli should be used to run server in production mode, not cube-server??? What about CUBEJS_DEV_MODE? Also, why not just use NODE_ENV which is the industry convention?
- Install cube-cli instead and try run the image: fails because there is no way (at least not documented) to specify where the cube.js config file is.

**Expected behavior**
The server should be able to be configured and run like any other node process. The pre-made docker image and project templates should be a convenience, not a requirement to hide the architectural decisions.

**Recommended solution**
The previous implementation of cube-server using a js entrypoint provided a better, more flexible DX in many situations. This should be a supported and recommended way to use cubejs for the forseeable future. 

To use:
- Install cube-server and required db connector.
- Create entry-point js file i.e. index.js which creates a cube-server instance.
- Run node index.js

**Comment by paveltiunov** [comment]
@MattGson Hey Matt! Thanks for posting such a detailed issue description! 

Could you please tell us a little bit more about your setup? Are you using Cube.js inside of the bigger Node.js app? If so what other parts of an app do? How do you deploy and scale it in production currently?

**Comment by MattGson** [comment]
@paveltiunov We are using Cube.js as part of a bigger SaaS app. It is deployed as its own docker image alongside several other node.js services. We have an [NX](https://nx.dev/angular) style mono-repo setup (single package dependency location) and use typescript with project references. 

The services have access to shared libs which include things like APM, ALM, Online debugging, Shared env configuration etc. This code is included in the modules for each service including the Cube.js server. (I have opened a separate issue for enabling all configuration through code to better support this). 

We build a base docker image from the mono-repo then package each service into smaller docker images. This allows us to add more services while maintaining a single cohesive build system. We scale horizontally on Kubernetes by adding more pods for each image as required.

This gives us a single cohesive build and configuration system and is a pretty standard setup for this kind of mono-repo.

**Comment by rpaik** [comment]
👋 @MattGson thanks for this issue. We're setting up a [Discourse forum](https://forum.cube.dev/). 

As a part of populating the initial content on the forum, we're looking into migrating some of the GH issues. We thought this issue is a good candidate for Discourse and hopefully, conversations will pick up there. (You'll be able to login to Discourse with your GitHub/Google/Twitter IDs). 

If you'd like, you can copy/paste your question on Discourse under the `Ideas` category by creating a "new topic". If you want us to take care of the post, we'll be happy to take care of it. LMK :slightly_smiling_face:

**Comment by rpaik** [comment]
🙇 @MattGson for opening a [Discourse discussion](https://forum.cube.dev/t/docker-architecture-provides-bad-dx-from-github-issues/142). Will close the issue here...
